proposals/0000-specialise-expressions.rst

```diff
@@ -0,0 +1,248 @@
+Allow expressions in SPECIALISE pragmas
+=======================================
+
+.. author:: Richard Eisenberg
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/493>`_.
+.. contents::
+.. sectnum::
+
+GHC today supports ``SPECIALISE`` pragmas (American spelling also accepted) that cause
+GHC to create a (partially) monomorphised version of a polymorphic function, often
+increasing efficiency::
+
+  addMult :: Num a => a -> a -> a
+  addMult x y = x * 2 + y
+
+  {-# SPECIALISE addMult :: Int -> Int -> Int #-}
+
+With the ``SPECIALISE`` pragma, GHC will produce ::
+
+  addMultInt :: Int -> Int -> Int
+  addMultInt x y = x * 2 + y
+  {-# RULES "addMult/Int" addMult = addMultInt #-}  -- NB: applies only at type Int
+
+Note that ``addMultInt`` uses no indirection in its use of addition and multiplication,
+unlike ``addMult``, which makes two calls to dictionary functions, confounding branch
+prediction and slowing down execution. The ``RULE`` means that GHC will replace
+all calls of ``addMult`` at ``Int`` to use the faster version.
+
+The syntax of today's ``SPECIALISE`` requires a function and its specialised type.
+
+This proposal suggests to generalise the syntax of ``SPECIALISE`` to allow an
+arbitrary expression, not just a function name and a type. Happily, this is backward
+compatible, because type signatures are expressions, too. This will allow, for example ::
```

Comment:
Not quite an arbitrary expression, right?
```suggestion
This proposal suggests to generalise the syntax of ``SPECIALISE`` to allow a
function application with an optional type signature, not just a function name and a type. Happily, this is backward
compatible. This will allow, for example ::
```

Reply:
yes!